container/openejb-core/src/main/java/org/apache/openejb/junit/jupiter/ApplicationComposerExtension.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.openejb.junit.jupiter;
+
+import org.apache.openejb.OpenEJBRuntimeException;
+import org.apache.openejb.testing.ApplicationComposers;
+import org.junit.jupiter.api.extension.*;
+
+public class ApplicationComposerExtension implements BeforeEachCallback, AfterEachCallback {
```

Comment:
A few open questions:

1. isnt it beforeall/afterall callbacks? (way faster)
2. shouldn't "delegate" be stored in ExtensionContext to ensure it is well scoped and not limited to this particular threading+scope which is one of the multiple junit5 cases
3. What about other test instances you can get from the context? should they contribute to the app model?
4. Should it get a "Single"/mono extension (started once per jvm as tomee embedded), it is generally accurate for apps written this way
5. (very minor) I wouldn't mention extension in the annotation "ApplicatioComposerExtension" name (maybe `@WithApplicationComposer` or alike)

hope it makes sense

Reply:
Wow. Super fast :+1:

1. I thought about it: Doesn't `ApplicationComposers` require a concrete test instance to inject values and stuff in `delegate.before(...)`?` The required test instance is - afaik - not available when `beforeAll(...)` is called. However, the creation of the `delegate` might be conducted in a `beforeAll(...) call.
2. This is a good idea. I think, it can be done via `context.getRoot().getStore(...)`  the `ExtensionContext`.
3. Honestly, I did not thought about this use-case. We may get additional test instances via `context.getRequiredTestInstances().getAllInstances()`. Wdyt?
4. Should be very similar to `SingleApplicationComposerRunner` - I will take a look at it.
5. Good idea. Will change it to `RunWith...` to be inline with the namings in `openejb-junit5-backward`.

Reply:
Thx. I updated things related to 2. and 3.

I have to think about 1. (because I think, that I didnt get it) and read / study the documentation related to `DynamicTests` in JUnit 5. Do you have an example of "splitting test definitions into multiple classes"? 

Might get you wrong, but did you refer to  `Test Instance Lifecycles` when talking about `@ApplicationComposer(mode = [PER_JVM|PER_ALL|PER_EACH])`  ? 

So we get smth like this, if the mode is specified?

- PER_JVM --> `SingleApplicationComposerExtension` (like TomEE embedded?)
- PER_ALL --> `@TestInstance(Lifecycle.PER_CLASS)` //  new test instance will be created once per test class.
- PER_EACH --> `@TestInstance(Lifecycle.PER_METHOD)` // a new test instance will be created for each test method, test factory method, or test template method (default in JUnit 4)

Reply:
Thanks for the details! 

> The split of an app composer app in multiple class is actually what you use ;). The constructor of ApplicationComposers takes N objects, just pass all test classes if you want a single deployment for all tests and enable each test to deploy part of the app.

Afaik the extension context only provides the related test class. So it would be something like

- Get inner classes from the test class
- Create instances of these inner test classes
- Supply the test class as well as the N instances of the inner test classes to  `ApplicationComposers` (like it is done in `AppComposerAsJUnitRuleWithReusableModulesTest` via JUnit 4 rules)  ?

> About the lifecycle it is not 1-1 with test instance lifecycle but the same idea for the container lifecycle (but you can have one test instance per test and one container instance for all tests of a single class when you want to reset another extension data/field instances for example). We can align on the default if nothing is set (so add to the enum an AUTO mode ;)) but must be configurable.

Thanks!  I think, I am now getting a better understanding of the idea. I also checked the JUnit 5 extension impl of meecrowave, which was really helpful.

Reply:
I made some progress based on your feedback, but I think it is still heavily **WIP** ;)  It should now honor the test instance lifecycle. The `mode` thing is missing yet. I have to think about the impl for the proposed `mode` for controlling the container instances created during tests.

Reply:
The current impl relies on the test's life cycle to start a container or perform related injections.

Atm the impl contains annotations and extensions for

- `RunWithSingleApplicationComposer`
- `RunWithApplicationComposer`

as well as a 

- `TomEEEmbeddedExtension`

From my understanding, the proposed `mode` is solely responsible to configure how (or when) the underlying container is started? I.e. one container to run all tests within a single test class (PER_ALL), one container for each test method of one test class (PER_EACH) and one container to run multiple test classes (PER_JVM), right?  Is this the same idea, which is currently conducted through different JUnit4 runners? 

From my understanding of the current JUnit implementation, the `ApplicationComposer` seems to be something similar to the PER_ALL and the `EmbeddedTomEERunner` / `SingleApplicationComposer` to the PER_JVM; I think PER_EACH (test method) isn't implemented in the JUnit 4 runners yet.

Probably I am missing something but I think, I do not have the whole process or related use-cases in mind ;) - Maybe you , @rmannibucau , can add some more ideas / thoughts on the `mode` idea?